### ISSUE TITLE
Banding issue on some inference cases

### ISSUE THREAD

**Issue by EmergentBehaviour** [MIC-DKFZ/nnUNet#3028]
Been using the ResEncoder plans recently (ResEncXL on an H100), and noticed that for some cases during inferencing / validation there will be banding / partial segmentation on certain regions / entire organs, looking as though the organ was segmented correctly but then chopped up as shown below.

I've tested this both with the nnUNet predictor and the initial argmax values and get the same thing. Suspecting this has something to do with sliding window / patching.

<img width="611" height="569" alt="Image" src="https://github.com/user-attachments/assets/df01698d-5d85-42c6-9c0f-8e6cc219dbcc" />

**Comment by FabianIsensee** [comment]
HI can you please give us a way to reproduce that? For example share a checkpoint and test case? The image can also be any publicly available image as long as it has the same issue. Please also tell us more about the training setup, in particular whether ignore label was used

**Comment by FabianIsensee** [comment]
Ping @EmergentBehaviour

**Comment by LFetty** [comment]
Hi I would like to ship in as I saw some similiar behaviour in some finetunings I did. Also for me the artifacts arise in sagittal direction. I can share a recent checkpoint together with a sample. The interesting part that it seems it can't even overfit the training data due to this artifact.

**Comment by FabianIsensee** [comment]
Please, do. I would like to see this :-) Is it just happening for the ResEncXL model or also others?

**Comment by LFetty** [comment]
here is a link to the checkpoint, plan and image/labels: [https://filesender.aco.net/?s=download&token=33d018f7-562a-4c42-a0ca-dd353ced6982](url)
I only use the M sized model as I find it efficient enough!

Some additional information:
-) I start with pretrained weights from a previous run which does not have this artifacts.
-) the new run only has 25 cases to train from.
-) I looked at the exported propability maps and it seems like it does learn the prober shape of the organs but for smaller organs it has this strips. This is not the case for larger organs.

If you need more information, happy to help out!

**Comment by FabianIsensee** [comment]
Thanks for sharing. I downloaded everything and can confirm that there is an issue!

<img width="884" height="367" alt="Image" src="https://github.com/user-attachments/assets/2cd096d4-64cf-48e2-aa38-fb020f82fe5e" />

Looking at your training logs through

```
a = torch.load('checkpoint_final.pth', weights_only=False)
print(a['logging']['dice_per_class_or_region'][-5:])
```

I can see that the classes have reportedly poor performance in training as well:

> [[np.float32(0.9199119),
  np.float32(0.006426694),
  np.float32(0.01681759),
  np.float32(0.04012027)],
 [np.float32(0.90204763),
  np.float32(0.029459072),
  np.float32(0.05259356),
  np.float32(0.03744051)],
 [np.float32(0.92708087),
  np.float32(0.14577886),
  np.float32(0.013014006),
  np.float32(0.042969737)],
 [np.float32(0.90912133),
  np.float32(0.08153451),
  np.float32(0.011585631),
  np.float32(0.056415185)],
 [np.float32(0.91824365),
  np.float32(0.07742175),
  np.float32(0.016140094),
  np.float32(0.041463416)]]

You can see how all but the liver class are in the 1-7 Dice points regime.

I am about to leave for vacation and will have more time when I return (beginning of June). If possible, could you share the training dataset with me? Without it I don't know whether I can fully diagnose the problem

**Comment by LFetty** [comment]
Yeah the performance is not good at all which was very as the pretrained model already performs very good! I suspected that the low number of patients were the problem but I'm not 100% convinced. The main point why is that is a fold all run, and that it performs that poor on the training is very strange.

Yes I can share the data privately. should I send you a download link per mail?
As you leave for vacation, is there anything I can try to debug the model properly so I could maybe already provide more information?

btw Thanks for the fast reponse! :)

**Comment by FabianIsensee** [comment]
When you go to our website you can find a contact form to drop me a message. I think this is the best way without leaking my email :-)

https://www.dkfz.de/en/medical-image-computing/our-team

Feel free to debug as you see fit, but it might be fastest if I am just going to do it once I am back. 
You are starting your training from a pretrained model? Does the same happen with a non-pretrained model?

**Comment by LFetty** [comment]
Send the data over the team contact form.

Yes, I did already a few test but didn't have the time to go deeper. I have to say I modified the plan slightly so the new data is not interpolated (1x1x4 mm --> 1x1x3 mm). 

However, the results are the some when:
-) trained from scratch,
-) with pretained weights, but lower lr and fewer epochs
-) identical plan, full training

hope that helps!

**Comment by FabianIsensee** [comment]
For anyone wondering, I am currently testing on @LFetty 's data to see whether I can reproduce the issue

**Comment by LFetty** [comment]
seems like the main issue was that my docker container was running an old version of nnUNet with torch==2.10+python3.12. The banding only happened only for a low number of cases and not with higher numbers. However, the issue seems to be resolved after updating to the latest master. 

I tested with python3.11+torch2.2.0 and with python 3.12+torch.2.10.0: both worked completely fine, no issues there.

One side note: 
I realized that the new pyproject.toml forces torch==2.2.0 if installed via `uv sync`. Needed to remove the macos_intel optional to make it install the higher torch version.

Bsides that thanks for the fast support and testing!

**Comment by FabianIsensee** [comment]
Claude: 

>   The intel_macos extra (added in a33e53d0) declares torch<2.3. When uv sync builds the lockfile, by
>   default it resolves a single set of versions that has to satisfy every extra simultaneously — even
>   ones you didn't ask for. The intersection of:
>   
>   - main torch>=2.1.2,!=2.9.*
>   - intel_macos torch<2.3
> 
>   …is >=2.1.2,<2.3, so the resolver picks the newest match, which is 2.2.2 (or 2.2.0 depending on
>   what's available). Removing the extra widens the window back to the latest 2.x.

Thanks uv, I guess...

I changed the  pyproject.toml now so that it should only apply those constraints when actually installing this on an intel system.
This is not the first issue with uv... pip is fine

**Comment by LFetty** [comment]
that unfortunetly does not fix the issue with sync. I guess if you want to support the sync command from uv it needs to separate the dependencies. That works for me:

```
[dependency-groups]
torch-uv= [
    "torch>=2.1.2,!=2.9.*",
]

[tool.uv]
conflicts = [
    [
        { group = "torch-uv" },
        { extra = "intel_macos" },
    ],
] 
```

then a user can run `uv sync --group torch-uv` to bypass that. Not really pretty to be honest...

**Comment by FabianIsensee** [comment]
OK I revert it and let uv users figure this out by themselves. This is a uv quirk that does not require fixing on our side. The recommended install path is pip :-)

**Comment by FabianIsensee** [comment]
Can I close this now?

**Comment by LFetty** [comment]
sorry for the late reply. from my side everything is clear and I guess the initial poster had the same issue. Thanks again!
